src/Symfony/Component/Translation/Translator.php

```diff
@@ -82,13 +77,28 @@ public function __construct(string $locale, MessageFormatterInterface $formatter
         $this->debug = $debug;
         $this->cacheVary = $cacheVary;
         $this->hasIntlFormatter = $formatter instanceof IntlFormatterInterface;
+        $this->fallbackLocaleProvider = $fallbackLocaleProvider ?? new FallbackLocaleProvider();
     }
 
     public function setConfigCacheFactory(ConfigCacheFactoryInterface $configCacheFactory)
     {
         $this->configCacheFactory = $configCacheFactory;
     }
 
+    public function setFallbackLocaleProvider(FallbackLocaleProviderInterface $fallbackLocaleProvider)
+    {
+        $this->fallbackLocaleProvider = $fallbackLocaleProvider;
+
+        // needed as the fallback locales are linked to the already loaded catalogues
+        $this->catalogues = [];
+        $this->cacheVary['fallback_locales'] = $fallbackLocaleProvider->getUltimateFallbackLocales();
```

Comment:
I don't like this for the following reasons:

When, in general, a `FallbackLocaleProvider` is responsible for giving the sequence of locales to try for a given starting locale, we should not assume it has this list of "last resort" locales, and we should also not assume that this configurable setting is the only thing affecting the computation results.

It might not be relevant in practice, but what if the JSON file with the locale hierarchy changes?

If the `FallbackLocaleProvider` just provides something like a cache key/suffix or similar (not exposing the `ultimateFallbackLocales` list), what would that mean for the "fallback locale" information gathered in the `DataCollectorTranslator`?

Reply:
I'm not sure to understand your concerns here. What you did seems really similar to what was done before `FallbackLocaleProvider` introduction.